src/librustc/traits/select.rs

```diff
@@ -2083,7 +2083,15 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                     }
 
                     if let Some(principal) = data.principal() {
-                        principal.with_self_ty(self.tcx(), self_ty)
+                        if !self.infcx.tcx.features().object_safe_for_dispatch {
+                            principal.with_self_ty(self.tcx(), self_ty)
+                        } else {
+                            if self.tcx().is_object_safe(principal.def_id()) {
+                                principal.with_self_ty(self.tcx(), self_ty)
+                            } else {
+                                return;
+                            }
+                        }
```

Comment:
Is the check for the feature flag necessary here? It seems like you could check if the trait is object safe anyway, even if the feature is not enabled. Or would that result in duplicate errors?

Reply:
I believe this is the code which adds the automatic 'impl Trait for dyn Trait' predicate to trait objects. Maybe the branches should be clarified and a comment added. Like:

// Add comments here
let object_impl_trait = !self.infcx.tcx.features().object_safe_for_dispatch ||
                                    self.tcx().is_object_safe(principal.def_id());

if (object_impl_trait) {
  principal.with_self_ty(self.tcx(), self_ty)
} else {
 return;
}

Reply:
At minimum, a comment would be good here -- I think I vaguely prefer the `if object_impl_trait { .. }` version as well.

But a comment like this would be great:

---

// Prior to RFC 2027, `dyn Foo: Foo` was always true. But RFC 2027 makes that only true if `Foo` is object safe.